cpp/include/cudf_test/stream_checking_resource_adaptor.hpp

```diff
@@ -131,25 +138,44 @@ class stream_checking_resource_adaptor final : public rmm::mr::device_memory_res
    */
   std::pair<std::size_t, std::size_t> do_get_mem_info(rmm::cuda_stream_view stream) const override
   {
-    verify_non_default_stream(stream);
+    verify_stream(stream);
     return upstream_->get_mem_info(stream);
   }
 
   /**
-   * @brief Throw an error if given one of CUDA's default stream specifiers.
+   * @brief Throw an error if the provided stream is invalid.
+   *
+   * A stream is invalid if:
+   * - check_default_stream_ is true and this function is passed one of CUDA's
+   *   default stream specifiers, or
+   * - check_default_stream_ is false and this function is passed any stream
+   *   other than the result of cudf::test::get_default_stream().
    *
-   * @throws `std::runtime_error` if provided a default stream
+   * @throws `std::runtime_error` if provided an invalid stream
    */
-  void verify_non_default_stream(rmm::cuda_stream_view const stream) const
+  void verify_stream(rmm::cuda_stream_view const stream) const
   {
     auto cstream{stream.value()};
-    if (cstream == cudaStreamDefault || (cstream == cudaStreamLegacy) ||
-        (cstream == cudaStreamPerThread)) {
-      throw std::runtime_error("Attempted to perform an operation on a default stream!");
+    auto const invalid_stream =
+      check_default_stream_ ? ((cstream == cudaStreamDefault) || (cstream == cudaStreamLegacy) ||
+                               (cstream == cudaStreamPerThread))
+                            : (cstream != cudf::test::get_default_stream().value());
+
+    if (invalid_stream) {
+      if (error_on_invalid_stream_) {
+        throw std::runtime_error("Attempted to perform an operation on an unexpected stream!");
+      } else {
+        std::cout << "Attempted to perform an operation on an unexpected stream!" << std::endl;
+      }
     }
   }
 
-  Upstream* upstream_;  // the upstream resource used for satisfying allocation requests
+  Upstream* upstream_;            // the upstream resource used for satisfying allocation requests
+  bool error_on_invalid_stream_;  // If true, throw an exception when the wrong stream is detected.
+                                  // If false, simply print to stdout.
+  bool check_default_stream_;  // If true, throw an exception when the default stream is observed.
+                               // If false, throw an exception when anything other than
+                               // cudf::test::get_default_stream() is observed.
```

Comment:
Our convention (and common usage in libcudf) is using an underscore as prefix, not suffix, for class private members.
```suggestion
  Upstream* _upstream;            // the upstream resource used for satisfying allocation requests
  bool _error_on_invalid_stream;  // If true, throw an exception when the wrong stream is detected.
                                  // If false, simply print to stdout.
  bool _check_default_stream;  // If true, throw an exception when the default stream is observed.
                               // If false, throw an exception when anything other than
                               // cudf::test::get_default_stream() is observed.
```

Reply:
That is true, but I chose to do it this way to follow the conventions of all rmm adaptors that I have seen (such as [the aligned adaptor](https://github.com/rapidsai/rmm/blob/branch-23.04/include/rmm/mr/device/aligned_resource_adaptor.hpp#L210) or [the logging adaptor](https://github.com/rapidsai/rmm/blob/branch-23.04/include/rmm/mr/device/logging_resource_adaptor.hpp#L304)). I can change it if you feel strongly about it, but I have a weak preference to leave it as is.